src/Storages/StorageMemory.cpp

```diff
@@ -41,29 +43,32 @@ class MemorySource : public SourceWithProgress
 protected:
     Chunk generate() override
     {
-        if (it == end)
+        if (is_finished)
         {
             return {};
         }
         else
         {
-            Block src = *it;
+            Block src = *first;
```

Comment:
We could easily save a couple of smaller memory allocations by not copying Block (it has at least std::vector + std::unordered_map). Just dereference the iterator and memorize the rows_count ... I hope this makes sense